rfcs/0000-compiled-runtime.md

```diff
@@ -0,0 +1,533 @@
+
+```jsx
+import { css, keyframes } from '@compiled/react';
+
+const fadeIn = keyframes({
+  to: { opacity: 1 },
+});
+
+const styles = css({
+  animationName: fadeIn,
+  animationDuration: '2s',
+});
+
+// 👇 transforms to
+
+const fadeIn = i([
+  '_j7hqa2t1',
+  ['._j7hqa2t1{animation-name:kfwl3rt}', '@keyframes kfwl3rt{to{opacity:1}}'],
+]);
+
+const styles = i([fadeIn, ['_5sagymdr', ['._5sagymdr{animation-duration:2s}']]]);
+```
+
+Breaks idiomatic CSS so the idea was tossed.
+
+**Considerations**
+
+- Could there be an alternate way to implement this that leans on similar patterns as the previous API?
+
+### CSS Prop
+
+CSS prop would now be wired up at runtime for both application of class names and style insertion. A component would be created for this: `CompiledElement`. This component would use the `ax` function to build a `className` as well as the `CC` and `CS` components to handle SSR/client side style insertion.
+
+```jsx
+/** @jsxImportSource @compiled/react */
+
+const redColorStyles = css({ color: 'red' });
+const hiddenStyles = css({ display: 'none' });
+const largeTextStyles = css({ fontSize: 50 });
+
+function Comp({ isRed, isHidden, isLarge }) {
+  return (
+    <div css={[isHidden ? hiddenStyles : [isLarge && largeTextStyles, isRed && redColorStyles]]} />
+  );
+}
+
+// 👇 transforms to
+
+const redColorStyles = i(['_syaz5scu', ['._syaz5scu{color:red}']]);
+const hiddenStyles = i(['_1e0cglyw', ['._1e0cglyw{display:none}']]);
+const largeTextStyles = i(['_1wyb1sen', ['._1wyb1sen{font-size:50}']]);
+
+function Comp({ isRed, isHidden, isLarge }) {
+  return (
+    <div css={[isHidden ? hiddenStyles : [isLarge && largeTextStyles, isRed && redColorStyles]]} />
+  );
+}
+```
+
+Style extraction then becomes a matter of pointing to the extract runtime and stripping out the styles from their arrays. The extract runtime would remove all style insertion behavior reducing the bundle size.
+
+```diff
+-/** @jsxImportSource @compiled/react */
++/** @jsxImportSource @compiled/react/extract */
+
+-const redColorStyles = i(['_syaz5scu', ['._syaz5scu{color:red}']]);
+-const hiddenStyles = i(['_1e0cglyw', ['._1e0cglyw{display:none}']]);
+-const largeTextStyles = i(['_1wyb1sen', ['._1wyb1sen{font-size:50}']]);
++const redColorStyles = ['_syaz5scu'];
++const hiddenStyles = ['_1e0cglyw'];
++const largeTextStyles = ['_1wyb1sen'];
+
+function Comp({ isRed, isHidden, isLarge }) {
+  return (
+    <div
+      css={[
+        isHidden
+          ? hiddenStyles
+          : [isLarge && largeTextStyles, isRed && redColorStyles],
+      ]}
+    />
+  );
+}
+```
+
+Less edge cases and AST transformation are needed to strip the runtime. Win win!
+
+### Styled
+
+Styled components would build on the same foundation as both the `css` func & prop, still leaning on a runtime to do the application of class names and style insertion.
```

Comment:
It would still be advantageous to block any non-DOM attributes from being passed down to the DOM element at runtime (related: https://github.com/atlassian-labs/compiled/issues/407)

This could look like another argument:

```jsx
styled.div`
  color: ${props => props.color};
`;

// to

styled('div', props => [
    ['_asd12s', ['_.asd12s{color:var(--as1112)}'], ['--as1112', props.color]]
  ], 
  ['color']); // <---
```